### PR TITLE
Timer IA: triggering on edge cases fixed. (#4089)

### LINKED ISSUE #4089: Timer: Stop triggering on edge cases
### Problem

We need to update the triggers for the Timer IA so it **doesn't trigger** under ambiguous cases. These are:

1. `20 minutes`
2. `60 minutes`
3. `48 hours`

_These are navigational queries where a user is trying to get to specific news websites_

screen shot 2017-04-13 at 3 28 59 pm

## Solution

1. Add 3 tests, each covering a case in [Timer.t](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/t/Timer.t)

2. Update [Timer.pm](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Timer.pm) to stop trigger on these specific edge cases.

3. Test it (`duckpan test timer`). Make sure all tests pass!

4. Create a PR

## People to notify

@pjhampton @moollaza 

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a Pull Request

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

---

Instant Answer Page: https://duck.co/ia/view/timer

### PR DESCRIPTION
**Pull request by shivam99aa**
## Description of new Instant Answer, or changes

It will stop invoking Timer IA on ambigous cases.

## Related Issues and Discussions

Fixes #4089 

## People to notify

@moollaza @pjhampton 

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/timer

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/Timer.pm [hunk 1]
```diff
@@ -104,6 +107,11 @@ handle remainder => sub {
         return;
     }
 
+    # When ambigous words are present in triggers then it should not
+    # invoke Timer IA.
+    if($raw =~ /(^|\s)($ambTrgx)(\s|$)/i) {
```
Comment (moollaza) [inline lib/DDG/Goodie/Timer.pm]: @shivam99aa this will prevent queries like `timer 20 minutes` which we want.

You want to use `/^$ambTrgx$/i`

---

Though it doesn't matter much here, we should probably replace `$->query_raw` with `->query_lc`. Raw is the original query without normalized spacing or letter casing. `query_lc` normalizes spacing and forces lowercase text meaning we don't need case sensitive regex. That would probably be a good follow-up optimization.

@mattr555 is there any reason we're using `query_raw` over `query` or `query_lc`?

File: lib/DDG/Goodie/Timer.pm [hunk 2]
```diff
@@ -18,6 +18,8 @@ my @beautifierTriggers = qw(online);
 my @joiners = qw(for on at with);
 # StartEndTriggers to trigger on nonStrippedTriggers, startTriggers, beautifierTriggers and triggers
 my @triggersStartEnd = (@triggers, @nonStrippedTriggers, @startTriggers, @beautifierTriggers);
+# Ambigous triggers which should not give Timer IA
+my @ambigousTriggers = ("20 minutes", "60 minutes", "48 hours");
```
Comment (moollaza) [inline lib/DDG/Goodie/Timer.pm]: @shivam99aa please add `22 minutes`, that's also the name of TV show here in Canada 👍

File: lib/DDG/Goodie/Timer.pm (hunk 1 above)
Comment (shivam99aa) [inline lib/DDG/Goodie/Timer.pm]: @moollaza Regex you are suggesting will not work for cases where user put a space at the start or end by mistake. For ex ` 20 minutes `. I propose to put `(^|^\s)$ambTrgx(\s$|$)` as regex.

File: lib/DDG/Goodie/Timer.pm (hunk 2 above)
Comment (shivam99aa) [inline lib/DDG/Goodie/Timer.pm]: Done in my code.

File: lib/DDG/Goodie/Timer.pm (hunk 1 above)
Comment (shivam99aa) [inline lib/DDG/Goodie/Timer.pm]: Even better would be to use `^s*$ambTrgx\s*$`. I will put this change and commit again.

**Comment by shivam99aa** [comment]
@moollaza I have done asked changes. Now it should stop Timer IA from coming in desired cases. Please review.

**Comment by mintsoft** [comment]
@moollaza this LGTM now it's updated

**Comment by moollaza** [comment]
@shivam99aa @mintsoft Thanks! This LGTM :+1:

File: lib/DDG/Goodie/Timer.pm (hunk 1 above)
Comment (moollaza) [inline lib/DDG/Goodie/Timer.pm]: FWIW this is why we should be using `query_lc`, or `query`. It normalizes spaces and trims the query so you don't have to deal with leading or trailing spaces.

@shivam99aa would you be interested in switching this to `query_lc`?

Comment (shivam99aa) [inline lib/DDG/Goodie/Timer.pm]: @moollaza Yeah I am ok with doing the change. Will you be creating a new issue or should I simply do the change and raise a pull request?

Comment (moollaza) [inline lib/DDG/Goodie/Timer.pm]: > Will you be creating a new issue or should I simply do the change and raise a pull request?

@shivam99aa just go ahead and make a PR if you'd like 👍 

Thanks!

Comment (shivam99aa) [inline lib/DDG/Goodie/Timer.pm]: @moollaza Sure I will make a PR for this change.
